_fcmaescpp/gcldeoptimizer.cpp

```diff
@@ -145,208 +149,219 @@ class GclDeOptimizer {
+		CR0 = CR0_;
+		// Number of iterations already performed.
+		iterations = 0;
+		bestValue = DBL_MAX;
+		// stop criteria
+		stop = 0;
+		//std::random_device rd;
+		rs = new pcg64(seed_);
+		init();
+	}
+
+	~GclDeOptimizer() {
+		delete rs;
+	}
+
+	double rnd01() {
+		return distr_01(*rs);
+	}
+
+	int rndInt(int max) {
+		return (int) (max * distr_01(*rs));
+	}
+
+	int findNear(vec x, vector<InfoStore> storage) {
+		int index = 0;
+		double minDist = DBL_MAX;
+		for (int j = 0; j < storage.size(); j++) {
+			double distance = (x - storage[j].x).squaredNorm();
+			if (distance < minDist) {
+				index = j;
+				minDist = distance;
+			}
+		}
+		return index;
+	}
+
+	void doOptimize() {
+		int Gr = 2;
+		int gen_stuck = 0;
+		vector<vec> sp;
+		vector<InfoStore> storage;
+
+		int maxIter = maxEvaluations / popsize + 1;
+		double previous_best = DBL_MAX;
+		double CR = CR0;
+		double F = F0;
+
+		// -------------------- Generation Loop --------------------------------
+
+		for (iterations = 1; fitfun->getEvaluations() < maxEvaluations;
+				iterations++) {
+			mat nextX = mat(popX);
+			vec nextY = vec(popY);
+			vec popCR = vec(popsize);
+			vec popF = vec(popsize);
+			vec xb = popX.col(0);
+			for (int p = 0; p < popsize; p++) {
+				vec xi = popX.col(p);
+				int r1, r2, r3;
+				do {
+					r1 = rndInt(popsize);
+				} while (r1 == p);
+				do {
+					r2 = rndInt(int(popsize * pbest));
+				} while (r2 == p || r2 == r1);
+				do {
+					r3 = rndInt(popsize + sp.size());
+				} while (r3 == p || r3 == r2 || r3 == r1);
+				int jr = rndInt(dim);
+				//Produce the CR and F
+				float mu = 1
+						- sqrt(float(iterations / maxIter))
+								* exp(float(-gen_stuck / iterations));
+				if (storage.empty() || iterations < Gr) {
+					CR = normreal(rs, 0.95, 0.01);
+					F = normreal(rs, mu, 1);
+					if (F < 0 || F > 1)
+						F = rnd01();
+				} else {
+					int i = findNear(popX.col(p), storage);
+					CR = storage[i].CR;
+					F = storage[i].F;
+				}
+				popCR[p] = CR;
+				popF[p] = F;
+				vec ui = vec(xi);
+				for (int j = 0; j < dim; j++) {
+					if (j == jr || rnd01() < CR) {
+						if (r3 < popsize)
+							ui[j] = popX(j, r1)
+									+ F * (popX(j, r2) - popX(j, r3));
+						else
+							ui[j] = popX(j, r1)
+									+ F * (popX(j, r2) - sp[r3 - popsize][j]);
+						if (ui[j] > 1 || ui[j] < 0)
+							ui[j] = rnd01();
+					}
+				}
+				nextX.col(p) = ui;
+			}
+			vec fitness = vec(popsize);
+			fitfun->values(nextX, popsize, fitness);
+			for (int p = 0; p < popsize; p++) {
 //            	fitness[p] = fitfun->value(nextX.col(p));
-                if (!isfinite(fitness[p]))
-                	fitness[p] = DBL_MAX;
-            }
-            for (int p = 0; p < popsize; p++) {
-                double y = fitness[p];
-                if (y < popY[p]) {
+				if (!isfinite(fitness[p]))
+					fitness[p] = DBL_MAX;
+			}
+			for (int p = 0; p < popsize; p++) {
+				double y = fitness[p];
+				if (y < popY[p]) {
 					nextY[p] = y;

```

Comment:
Please check here. The parent vector seems not replaced by the better child vector.

Reply:
Checked again, in
				nextX.col(p) = ui;
I set the child by the new candidate for the while population. 

Reason is that I later want 

fitfun->values(nextX, popsize, fitness);

to be able to evaluat the whole popsize in parallel, so  nextX should contain all candidates even the bad ones - 
which we find out after evaluation. 

This means, that in case there is no improvement, I have to revert the assignment and
switch nextX.col(p) back to the (better) parent.